### ISSUE TITLE
Error with Cosine Proximity as target for loss function (with merge layer)

### ISSUE THREAD

**Issue by pchowdhry** [fchollet/keras#5046]
Hi I'm trying to build a simple LSTM for a sentence similarity task. I'm using Glove word embeddings going into the LSTMs, and padding my sentences to 50. I then go into a merge layer on axis 1 using 'cos'. Whenever I fit and compile I get a nan for loss, immediately. I've tried to keep it simple for easy debugging. Any help would be appreciated.  Model summary and code below (left out data preprocessing)

total training data = 1184
X1.shape = (1184, 50)
X2.shape = (1184, 50)
Y.shape = (1184, 1)
____________________________________________________________________________________________________
Layer (type)                     Output Shape          Param #     Connected to                     
____________________________________________________________________________________________________
embedding_1 (Embedding)          (None, 50, 100)       310200      embedding_input_1[0][0]          
____________________________________________________________________________________________________
lstm_1 (LSTM)                    (None, 128)           117248      embedding_1[0][0]                
____________________________________________________________________________________________________
embedding_2 (Embedding)          (None, 50, 100)       310200      embedding_input_2[0][0]          
____________________________________________________________________________________________________
lstm_2 (LSTM)                    (None, 128)           117248      embedding_2[0][0]                
____________________________________________________________________________________________________
activation_1 (Activation)        (None, 1)             0           merge_1[0][0]                    
____________________________________________________________________________________________________
Total params: 854,896
Trainable params: 234,496
Non-trainable params: 620,400

```
embedding_layer_a = Embedding(len(word_index) + 1,
                            EMBEDDING_DIM,
                            weights=[embedding_matrix],
                            input_length=50,
                            trainable=False)

embedding_layer_b = Embedding(len(word_index) + 1,
                            EMBEDDING_DIM,
                            weights=[embedding_matrix],
                            input_length=50,
                            trainable=False)

s1rnn = Sequential()
s1rnn.add(embedding_layer_a)
s1rnn.add(LSTM(128, input_shape=(100, 1), dropout_W=0.2, dropout_U=0.2))


s2rnn = Sequential()
s2rnn.add(embedding_layer_b)
s2rnn.add(LSTM(128, input_shape=(100, 1), dropout_W=0.2, dropout_U=0.2))


model = Sequential()
merged = Merge([s1rnn, s2rnn], mode='cos', dot_axes=1)
model.add(merged)
model.add(Activation('relu'))

model.summary()

model.compile(optimizer='rmsprop', loss='cosine_proximity', metrics=['accuracy'])

model.fit([X1, X2], Y, batch_size=32, nb_epoch=500, validation_split=0.10)

```

**Comment by mbotumbo** [comment]
The cosine_proximity loss computes the cosine proximity between the output of the network and the _y_ values you supply to the fit function. Thus, with the setup you have here, it's computing the cosine between the output of your cosine-merge->relu with whatever your Y is. 

I think one option to do what it seems you are trying to do is have is a custom loss function that will result in minimizing the negative output of your network (and either disregards Y or uses Y as some margin-like value). 

As a side note, if your inputs are sentences from the same language, consider sharing your network that computes the representation using a network that is shared across your two inputs (e.g. siamese network).
